src/core/net/dnssd_server.cpp

```diff
@@ -0,0 +1,659 @@
+}
+
+void Server::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    otError  error = OT_ERROR_NONE;
+    Header   requestHeader;
+    Message *responseMessage = nullptr;
+
+    // Make sure DNS header is at offset zero to simplify server logic
+    aMessage.RemoveHeader(aMessage.GetOffset());
+
+    SuccessOrExit(error = aMessage.Read(0, requestHeader));
+    VerifyOrExit(requestHeader.GetType() == Header::kTypeQuery, error = OT_ERROR_DROP);
+
+    responseMessage = NewMessage();
+    VerifyOrExit(responseMessage != nullptr, error = OT_ERROR_NO_BUFS);
+
+    // Allocate space for DNS header and move offset after DNS header
+    SuccessOrExit(error = responseMessage->SetLength(sizeof(Header)));
+
+    ProcessQuery(aMessage, *responseMessage, requestHeader);
+
+    error = SendMessage(*responseMessage, aMessageInfo);
+
+exit:
+    FreeMessageOnError(responseMessage, error);
+}
+
+void Server::ProcessQuery(Message &aMessage, Message &aResponse, const Header &aRequestHeader)
+{
+    Header           responseHeader;
+    uint16_t         readOffset, nameSerializeOffset;
+    Question         question;
+    uint16_t         qtype;
+    char             name[Dns::Name::kMaxLength + 1];
+    otError          error = OT_ERROR_NONE;
+    NameCompressInfo compressInfo;
+
+    // Setup initial DNS response header
+    responseHeader.Clear();
+    responseHeader.SetResponseCode(Header::kResponseSuccess);
+    responseHeader.SetType(Header::kTypeResponse);
+    responseHeader.SetMessageId(aRequestHeader.GetMessageId());
+
+    // Validate the query
+    VerifyOrExit(aRequestHeader.GetQueryType() == Header::kQueryTypeStandard,
+                 responseHeader.SetResponseCode(Header::kResponseNotImplemented));
+    VerifyOrExit(!aRequestHeader.IsTruncationFlagSet(), responseHeader.SetResponseCode(Header::kResponseFormatError));
+    VerifyOrExit(aRequestHeader.GetQuestionCount() == 1, responseHeader.SetResponseCode(Header::kResponseFormatError));
+
+    // Read the query name and question
+    readOffset = sizeof(Header);
+    VerifyOrExit(OT_ERROR_NONE == Dns::Name::ReadName(aMessage, readOffset, name, sizeof(name)),
+                 responseHeader.SetResponseCode(Header::kResponseFormatError));
+    VerifyOrExit(OT_ERROR_NONE == aMessage.Read(readOffset, question),
+                 responseHeader.SetResponseCode(Header::kResponseFormatError));
+
+    // Add the question to the response, and save the serialize offset of name
+    nameSerializeOffset = aResponse.GetLength();
+    SuccessOrExit(error = AddQuestionToResponse(name, question, aResponse, responseHeader));
+
+    // Further validate the query
+    qtype = question.GetType();
+    VerifyOrExit(qtype == ResourceRecord::kTypePtr || qtype == ResourceRecord::kTypeSrv ||
+                     qtype == ResourceRecord::kTypeTxt || qtype == ResourceRecord::kTypeAaaa,
+                 responseHeader.SetResponseCode(Header::kResponseNotImplemented));
```

Comment:
I was thinking about it but didn't do this way because of the calling to `ResolveQuestion` :
```c++
    // Resolve the question
    SuccessOrExit(error = ResolveQuestion(name, question, responseHeader, aResponse, compressInfo));
```

`ResolveQuestion` could set the RCODE and currently it just modify the `responseHeader`. If we define a new variable, we would need to add a out parameter `aResponseCode` to the `ResolveQuestion`. 

I am not sure which is simpler. I will revise it later with further experiments.